File: algorithms/11_golang/math/ncr.go
package main

import "fmt"

const mod = 1_000_000_007

func modPow(a, b, m int) int {
	a %= m
	res := 1
	for b > 0 {
		if b&1 == 1 {
			res = res * a % m
		}
		a = a * a % m
		b >>= 1
	}
	return res
}

var fact, invFact []int

func initFact(n int) {
	fact = make([]int, n+1)
	invFact = make([]int, n+1)
	fact[0] = 1
	for i := 1; i <= n; i++ {
		fact[i] = fact[i-1] * i % mod
	}
	invFact[n] = modPow(fact[n], mod-2, mod)
	for i := n - 1; i >= 0; i-- {
		invFact[i] = invFact[i+1] * (i + 1) % mod
	}
}

func nCr(n, r int) int {
	if r < 0 || r > n {
		return 0
	}
	return fact[n] * invFact[r] % mod * invFact[n-r] % mod
}

func main() {
	initFact(100)
	if nCr(5, 2) != 10 || nCr(10, 3) != 120 || nCr(10, 10) != 1 {
		panic("nCr mod p failed")
	}
	fmt.Println("[Go nCr] Binomial coefficients mod p verified.")
}

File: algorithms/06_javascript/math/ncr.js
function mulMod(a, b, p) {
  return Number((BigInt(a) * BigInt(b)) % BigInt(p));
}

function modPow(base, exp, mod) {
  let result = 1;
  base = base % mod;
  while (exp > 0) {
    if (exp & 1) result = mulMod(result, base, mod);
    base = mulMod(base, base, mod);
    exp >>= 1;
  }
  return result;
}

function nCrModP(n, r, p) {
  if (r < 0 || r > n) return 0;
  const fact = new Array(n + 1);
  const invFact = new Array(n + 1);
  fact[0] = 1;
  for (let i = 1; i <= n; i++) fact[i] = mulMod(fact[i - 1], i, p);
  invFact[n] = modPow(fact[n], p - 2, p);
  for (let i = n; i >= 1; i--) invFact[i - 1] = mulMod(invFact[i], i, p);
  return mulMod(mulMod(fact[n], invFact[r], p), invFact[n - r], p);
}

module.exports = { nCrModP, modPow };

if (require.main === module) {
  if (nCrModP(5, 2, 1000000007) !== 10) throw new Error("nCr failed");
  if (nCrModP(10, 3, 1000000007) !== 120) throw new Error("nCr failed");
  if (nCrModP(10, 12, 1000000007) !== 0) throw new Error("nCr invalid range failed");
  console.log("[JavaScript nCr] Combinations mod p verified");
}
